src/Workspaces/Core/Portable/Shared/Extensions/SyntaxTokenListExtensions.cs

```diff
@@ -1,12 +1,20 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System.Collections.Generic;
+using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.Shared.Extensions
 {
     internal static class SyntaxTokenListExtensions
     {
         public static SyntaxTokenList ToSyntaxTokenList(this IEnumerable<SyntaxToken> tokens)
             => new SyntaxTokenList(tokens);
+
+        public static SyntaxTokenList ToSyntaxTokenListAndFree(this ArrayBuilder<SyntaxToken> tokens)
+        {
+            var tokenList = new SyntaxTokenList(tokens);
+            tokens.Free();
+            return tokenList;
```

Comment:
It woudl be great if there were a SyntaxTokenList's constructor that took an ArrayBuilder.  That way we could enumerate it without allocating.   RIght now, this will call the IEnumerable constructor (i think), which will allocate the IEnumerator to walk the tokens.

Reply:
Tracking in #34960